Custom styles

eginhard opened the pull request:
Are you happy to take pull requests here?

This provides a mechanism to supply a custom pybtex style via the Pelican config, so that the display of the publication list can be customised without touching this plugin's code. I added a basic example to the README.

eginhard committed:
Update .gitignore

eginhard committed:
Refactor.

eginhard committed:
Allow supplying custom pybtex styles.

eginhard committed:
Add custom style example to README.

scheunemann commented:
Of course PRs are welcome, thank you for that :+1: The refactoring looks cool and I will test the plugin-functionality when I am back at my computer.

One question popped up immediately: I assume you picked that example (which is probably the most popular use case) for the sake of illustration. However, is there a benefit of using the plugin over `jinja` filters like, e.g., `{{ publication.bibtex | replace("author_name","<strong>author_name</strong>")  }}` ?

(not tested, typed from memory)

eginhard commented:
Yes, something like that might be simpler in this case. It's probably more useful for changing the general ordering and layout of the fields. Here is another style that you can try:

```python
from pybtex.style.formatting import toplevel, unsrt
from pybtex.style.template import (
    field, first_of, href, join, names, optional, optional_field, sentence,
    tag, together, words
)

class WebStyle(unsrt.Style):
    def get_article_template(self, e):
        volume = join [
            field('volume'),
            optional['(', field('number'),')']
        ]
        template = toplevel [
            self.format_title(e, 'title'),
            self.format_names('author'),
            sentence [
                tag('em') [field('journal')],
                optional[ volume ],
            ],
            sentence [ field('year') ],
            self.format_note(e),
        ]
        return template

    def get_inproceedings_template(self, e):
        template = toplevel [
            self.format_title(e, 'title'),
            sentence [self.format_names('author')],
            words [
                'In',
                sentence [
                    self.format_btitle(e, 'booktitle', as_sentence=False),
                    self.format_volume_and_series(e, as_sentence=False),
                ],
                sentence [ field('year') ],
            ],
            self.format_note(e),
        ]
        return template

    def format_note(self, e):
        return optional [tag('mark') [
            words(sep='') [ '(', sentence [ field('note') ] , ')' ]
        ]]

    def format_title(self, e, which_field, as_sentence=True):
        title = tag('strong') [ field(
            which_field, apply_func=lambda text: text.capitalize()
        )]
        formatted_title = first_of [
            optional [
                href [
                    field('url', raw=True),
                    title
                ]
            ],
            title
        ]
        if as_sentence:
            return sentence [ formatted_title ]
        else:
            return formatted_title
```

scheunemann commented:
@eginhard first of all, that is a really neat feature and I am very happy about your PR. Cool cool! I had a look and tested everything, it works all fine and I can see it being very helpful, I personally like the webstyle idea from above a lot!.

One idea from my side though. I think having python code in the `pelicanconf.py` is okay, but it could be more tidy without. What do think about the following:

## option 1
The users can specify a path to a file called `bibstyle.py` with `PUBLICATIONS_STYLE_PATH`, alternatively they have to drop the file into `plugins` as default. If the user specifies `PUBLICATIONS_CUSTOM_STYLE = True`, we automatically use the style of that file?

## option 2
We let the user point via `PUBLICATIONS_PLUGIN_PATH` to a folder with several custom styles which we automatically import. Then the variable `PUBLICATIONS_STYLE` can be assigned with one of these styles (that can cause some nasty clashes with other plugins though).

In both ways, the `pelicanconf.py` stays clean. Option 1 feels most natural, as we can have only one custom style anyway, so why allowing the freedom for having several classes with different names?

What are your thoughts here?

eginhard committed:
Handle imports in bib.py

eginhard commented:
Thanks for the feedback, that sounds good. I added a commit that implements option 1. Now it's looking for `pybtex_plugins.py` in `PUBLICATIONS_STYLE_PATH` from which it tries to import a `PelicanStyle`.

I suggest to use the more general `pybtex_plugins.py` because it can allow to customise other parts of pybtex in a similar way. For example, I added a custom bibtex writer that adds linebreaks to long lines in the bibtex output on another branch: https://github.com/eginhard/pelican-bib/commit/29b26877ac4fc8fa75cdad9c77a3fc3c638ec7aa

That is used with `PUBLICATIONS_CUSTOM_BIBTEX_WRITER = True` and the following plugin:
```python
import textwrap
from pybtex.database.output.bibtex import Writer

class PelicanBibtexWriter(Writer):
    def _write_field(self, stream, field, value):
        value = self.quote(self._encode(value))
        subsequent = ' ' * (len(f'    {field} = "'))
        lines = textwrap.wrap(f'{field} = {value}', initial_indent='    ',
                              subsequent_indent=subsequent)
        stream.write(',')
        for line in lines:
            stream.write(f'\n{line}')
```

scheunemann commented:
Thank you @eginhard ! That's looking good to me, only the [`README.md`](https://github.com/eginhard/pelican-bib/blob/custom-styles/README.md) will need some updating again.

I will take care of testing asap.

eginhard committed:
Update README

eginhard commented:
I updated the README.

scheunemann commented:
@eginhard that looks all good to me. Thank you again for the cool contribution.

I will merge that, but won't push it to PyPI just now, if that is okay with you. I will first think about the following:

1. Variable naming: We could, for example, check whether the file `pybtex_plugins.py` exists, and then do change the style if `PelicanStyle()` is defined (or maybe later, `PelicanBibtexWriter()`), rather than asking the user to define `PUBLICATIONS_CUSTOM_STYLE = True`. For some reason, the variable feels redundant now.
2. *Plugins for plugins*: I had a scroll through other pelican plugins, but I couldn't find an example of an extendable plugin. The process here feels alright to me though, but I want to give it a further look if we are missing a *standard way*.

Happy for any further input, I am now looking forward to style my own publications list wayyyyy nicer! :+1:

eginhard commented:
Thanks!
1. I actually thought about doing that. The current way allows slightly more fine-grained error messages depending on whether the module or the class is not found, but I agree that it's probably not necessary.
2. I'm just getting started with Pelican, so I'm not familiar with any best practices in that regard either, unfortunately. I'll see if I find anything.
